File: provider/src/main/java/com/raygun/raygun4android/RaygunSettings.java
package com.raygun.raygun4android;

import java.util.Arrays;
import java.util.HashSet;

public class RaygunSettings {

    // General
    public static final String APIKEY_MANIFEST_FIELD = "com.raygun.raygun4android.apikey";
    public static final String RAYGUN_CLIENT_VERSION = BuildConfig.VERSION_NAME;

    // HTTP error response codes
    public static final int RESPONSE_CODE_ACCEPTED = 202;
    public static final int RESPONSE_CODE_BAD_MESSAGE = 400;
    public static final int RESPONSE_CODE_INVALID_API_KEY = 403;
    public static final int RESPONSE_CODE_LARGE_PAYLOAD = 413;
    public static final int RESPONSE_CODE_RATE_LIMITED = 429;

    // Crash Reporting
    public static final String DEFAULT_CRASHREPORTING_ENDPOINT = "https://api.raygun.io/entries";
    public static final String DEFAULT_FILE_EXTENSION = "raygun4";
    public static final int DEFAULT_MAX_REPORTS_STORED_ON_DEVICE = 64;
    public static final String CRASH_REPORTING_UNHANDLED_EXCEPTION_TAG = "UnhandledException";

    // RUM
    public static final String RUM_EVENT_SESSION_START = "session_start";
    public static final String RUM_EVENT_SESSION_END = "session_end";
    public static final String RUM_EVENT_TIMING = "mobile_event_timing";
    public static final String DEFAULT_RUM_ENDPOINT = "https://api.raygun.io/events";
    // 30 minutes in milliseconds
    public static final int RUM_SESSION_EXPIRY = 30 * 60 * 1000;

    private static IgnoredURLs ignoredURLs = new IgnoredURLs("api.raygun.io");
    private static HashSet<String> ignoredViews = new HashSet<>();
    private static int maxReportsStoredOnDevice = DEFAULT_MAX_REPORTS_STORED_ON_DEVICE;
    private static String crashReportingEndpoint = DEFAULT_CRASHREPORTING_ENDPOINT;
    private static String rumEndpoint = DEFAULT_RUM_ENDPOINT;

    private RaygunSettings() {
    }

    public static String getCrashReportingEndpoint() {
        return crashReportingEndpoint;
    }

    static void setCrashReportingEndpoint(String crashReportingEndpoint) {
        RaygunSettings.crashReportingEndpoint = crashReportingEndpoint;
    }

    public static String getRUMEndpoint() {
        return rumEndpoint;
    }

    static void setRUMEndpoint(String rumEndpoint) {
        RaygunSettings.rumEndpoint = rumEndpoint;
    }

    public static HashSet<String> getIgnoredURLs() {
        return ignoredURLs;
    }

    static HashSet<String> getIgnoredViews() {
        return ignoredViews;
    }

    public static int getMaxReportsStoredOnDevice() {
        return maxReportsStoredOnDevice;
    }

    static void setMaxReportsStoredOnDevice(int maxReportsStoredOnDevice) {
        if (maxReportsStoredOnDevice <= DEFAULT_MAX_REPORTS_STORED_ON_DEVICE) {
            RaygunSettings.maxReportsStoredOnDevice = maxReportsStoredOnDevice;
        } else {
            RaygunLogger.w("It's not possible to exceed the value " + DEFAULT_MAX_REPORTS_STORED_ON_DEVICE + " for the number of reports stored on the device. The setting has not been applied.");
        }
    }

    public static class IgnoredURLs extends HashSet<String> {
        IgnoredURLs(String... defaultIgnoredUrls) {
            this.addAll(Arrays.asList(defaultIgnoredUrls));
        }
    }

    static void ignoreURLs(String[] urls) {
        if (urls != null) {
            for (String url : urls) {
                if (url != null) {
                    RaygunSettings.ignoredURLs.add(url);
                }
            }
        }
    }

    static void ignoreViews(String[] views) {
        if (views != null) {
            for (String view : views) {
                if (view != null) {
                    RaygunSettings.ignoredViews.add(view);
                }
            }
        }
    }
}

File: provider/src/main/java/com/raygun/raygun4android/RaygunSettings.kt
@file:Suppress("MemberVisibilityCanBePrivate")

package com.raygun.raygun4android

import com.raygun.raygun4android.logging.RaygunLogger.w
import com.raygun.raygun4android.network.RaygunOkHttpClientBuilder
import okhttp3.OkHttpClient

object RaygunSettings {
    // General
    const val APIKEY_MANIFEST_FIELD: String = "com.raygun.raygun4android.apikey"
    const val RAYGUN_CLIENT_VERSION: String = BuildConfig.VERSION_NAME
    const val LOGGING_TAG: String = "Raygun4Android"

    // HTTP error response codes
    const val RESPONSE_CODE_ACCEPTED: Int = 202
    const val RESPONSE_CODE_BAD_MESSAGE: Int = 400
    const val RESPONSE_CODE_INVALID_API_KEY: Int = 403
    const val RESPONSE_CODE_LARGE_PAYLOAD: Int = 413
    const val RESPONSE_CODE_RATE_LIMITED: Int = 429

    // Crash Reporting
    const val DEFAULT_CRASHREPORTING_ENDPOINT: String = "https://api.raygun.io/entries"
    const val DEFAULT_FILE_EXTENSION: String = "raygun4"
    const val DEFAULT_MAX_REPORTS_STORED_ON_DEVICE: Int = 64
    const val CRASH_REPORTING_UNHANDLED_EXCEPTION_TAG: String = "UnhandledException"

    // RUM
    const val RUM_EVENT_SESSION_START: String = "session_start"
    const val RUM_EVENT_SESSION_END: String = "session_end"
    const val RUM_EVENT_TIMING: String = "mobile_event_timing"
    const val DEFAULT_RUM_ENDPOINT: String = "https://api.raygun.io/events"

    // 30 minutes in milliseconds
    const val RUM_SESSION_EXPIRY: Int = 30 * 60 * 1000

    private val ignoredURLs = IgnoredURLs("api.raygun.io")

    val ignoredViews: HashSet<String> = HashSet()

    @JvmStatic
    var maxReportsStoredOnDevice: Int = DEFAULT_MAX_REPORTS_STORED_ON_DEVICE
        set(maxReportsStoredOnDevice) {
            if (maxReportsStoredOnDevice <= DEFAULT_MAX_REPORTS_STORED_ON_DEVICE) {
                field = maxReportsStoredOnDevice
            } else {
                w(
                    (
                        "It's not possible to exceed the value " +
                            DEFAULT_MAX_REPORTS_STORED_ON_DEVICE +
                            " for the number of reports stored on the device. The setting has not" +
                            " been applied."
                    ),
                )
            }
        }

    @JvmField var crashReportingEndpoint: String = DEFAULT_CRASHREPORTING_ENDPOINT

    var rumEndpoint: String = DEFAULT_RUM_ENDPOINT

    @JvmField var okHttpClientBuilder: OkHttpClientBuilder? = null

    fun getIgnoredURLs(): HashSet<String> = ignoredURLs

    fun ignoreURLs(urls: Array<String?>?) {
        if (urls != null) {
            for (url in urls) {
                if (url != null) {
                    ignoredURLs.add(url)
                }
            }
        }
    }

    fun ignoreViews(views: Array<String?>?) {
        if (views != null) {
            for (view in views) {
                if (view != null) {
                    ignoredViews.add(view)
                }
            }
        }
    }

    val httpClient: OkHttpClient
        get() {
            return if (okHttpClientBuilder != null) {
                okHttpClientBuilder!!.build()
            } else {
                RaygunOkHttpClientBuilder.instance.build()
            }
        }

    class IgnoredURLs internal constructor(
        vararg defaultIgnoredUrls: String,
    ) : HashSet<String>() {
        init {
            this.addAll(defaultIgnoredUrls.toList())
        }
    }
}
